dev/tools/localizations_utils.dart

```diff
@@ -9,6 +9,136 @@ import 'dart:io';
 import 'package:args/args.dart' as argslib;
 import 'package:meta/meta.dart';
 
+/// Simple data class to hold parsed locale. Does not promise validity of any data.
+class LocaleInfo implements Comparable<LocaleInfo> {
+  LocaleInfo({
+    this.languageCode,
+    this.scriptCode,
+    this.countryCode,
+    this.length,
+    this.originalString,
+  });
+
+  /// Simple parser. Expects the locale string to be in the form of 'language_script_COUNTRY'
+  /// where the langauge is 2 characters, script is 4 characters with the first uppercase,
+  /// and country is 2-3 characters and all uppercase.
+  ///
+  /// 'language_COUNTRY' or 'language_script' are also valid. Missing fields will be null.
+  factory LocaleInfo.fromString(String locale, {bool assume = false}) {
+    final List<String> codes = locale.split('_'); // [language, script, country]
+    assert(codes.isNotEmpty && codes.length < 4);
+    final String languageCode = codes[0];
+    String scriptCode;
+    String countryCode;
+    int length = codes.length;
+    String originalString = locale;
+    if (codes.length == 2) {
+      scriptCode = codes[1].length >= 4 ? codes[1] : null;
+      countryCode = codes[1].length < 4 ? codes[1] : null;
+    } else if (codes.length == 3) {
+      scriptCode = codes[1].length > codes[2].length ? codes[1] : codes[2];
+      countryCode = codes[1].length < codes[2].length ? codes[1] : codes[2];
+    }
+    assert(codes[0] != null && codes[0].isNotEmpty);
+    assert(countryCode == null || countryCode.isNotEmpty);
+    assert(scriptCode == null || scriptCode.isNotEmpty);
+
+    /// Adds scriptCodes to locales where we are able to assume it to provide
+    /// finer granularity when resolving locales.
+    ///
+    /// The basis of the assumptions here are based off of known usage of scripts
+    /// across various countries. For example, we know Taiwan uses traditional (Hant)
+    /// script, so it is safe to apply (Hant) to Taiwanese languages.
+    if (assume && scriptCode == null) {
+      switch (languageCode) {
+        case 'zh': {
+          switch (countryCode) {
+            case 'CN':
+              scriptCode = 'Hans';
+              break;
+            case 'SG':
+              scriptCode = 'Hans';
+              break;
+            case 'TW':
+              scriptCode = 'Hant';
+              break;
+            case 'HK':
+              scriptCode = 'Hant';
+              break;
+            case 'MO':
+              scriptCode = 'Hant';
+              break;
+          }
+          if (countryCode == null) {
```

Comment:
Since a null countryCode will not match a case, could put it ahead of the switch